### PR TITLE
Add symint support for select, unfold, unfold_backward

### PR DESCRIPTION
**Pull request by anjali411**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #87230

### CODE REVIEW

File: aten/src/ATen/native/ComplexHelper.h [hunk 1]
```diff
@@ -33,13 +33,13 @@ inline DimVector computeStrideForViewAsReal(IntArrayRef oldstride) {
...
   TORCH_CHECK(self.is_complex(), "view_as_real is only supported for complex tensors");
-  auto old_sizes = self.sizes();
-  DimVector new_sizes(old_sizes.size() + 1);
+  auto old_sizes = self.sym_sizes();
+  SymDimVector new_sizes(old_sizes.size() + 1);
   std::copy(old_sizes.begin(), old_sizes.end(), new_sizes.begin());
   // last dimension will always have two elements containing the real and imag vals
   new_sizes.back() = 2;
-  auto new_strides = computeStrideForViewAsReal(self.strides());
-  auto new_storage_offset = 2 * self.storage_offset();
```
Comment (ezyang) [inline aten/src/ATen/native/ComplexHelper.h:42]: fwiw this should work

File: aten/src/ATen/native/TensorShape.cpp [hunk 2]
```diff
@@ -3374,10 +3374,10 @@ Tensor unfold(const Tensor& self, int64_t d, int64_t size, int64_t step) {
   auto ndim = self.dim();
   d = at::maybe_wrap_dim(d, ndim, /*wrap_scalar=*/true);
 
-  auto sizes = self.sizes().vec();
-  auto strides = self.strides().vec();
-  int64_t max_size = self.dim() == 0 ? 1 : sizes[d];
-  TORCH_CHECK(size <= max_size, "maximum size for tensor at dimension ", d,
```
Comment (ezyang) [inline aten/src/ATen/native/TensorShape.cpp:3380]: this should work

File: c10/core/TensorImpl.cpp [hunk 3]
```diff
@@ -990,6 +990,12 @@ void TensorImpl::set_sizes_and_strides(
   refresh_contiguous();
 }
 
+void TensorImpl::set_storage_offset(c10::SymInt storage_offset) {
+  if (!extra_meta_) {
+    extra_meta_ = std::make_unique<ExtraMeta>();
+  }
+  extra_meta_->storage_offset_ = storage_offset.clone();
+}
```
Comment (ezyang) [inline c10/core/TensorImpl.cpp:998]: No, this is not sound. set_sizes_and_strides has a bunch of logic for routing to the normal non `extra_meta_` case when there are no symbolics anywhere, and it needs to be replicated here

**Review by ezyang — CHANGES_REQUESTED** [review]
set_storage_offset is not right

File: c10/core/TensorImpl.cpp (hunk 3 above)
Comment (ezyang) [reply c10/core/TensorImpl.cpp:998]: or better yet, we continue to use `set_sizes_and_strides` so we only need one implementation of these functions

**Comment by anjali411** [comment]
> set_storage_offset is not right

yup removed it. it's not necessary

**Review by ezyang — APPROVED** [review]
thanks!

File: aten/src/ATen/functorch/BatchRulesDecompositions.cpp [hunk 4]
```diff
@@ -262,8 +262,6 @@ TORCH_LIBRARY_IMPL(aten, FuncTorchBatched, m) {
   OP_DECOMPOSE(t_);
   OP_DECOMPOSE(swapdims_);
   OP_DECOMPOSE(swapaxes_);
-  OP_DECOMPOSE(unfold_copy);
```
Comment (anjali411) [inline aten/src/ATen/functorch/BatchRulesDecompositions.cpp:265]: not a composite implicit autograd op, so shouldn't be here 

(This was caught through a segfault at test_vjpvmap_unfold_copy_cpu_float32)

**Review by ezyang — APPROVED** [review]
reapproving without review

**Comment by anjali411** [comment]
```
ERROR [0.013s]: test_op_has_batch_rule_unfold_copy_cpu_float32 (__main__.TestVmapOperatorsOpInfoCPU)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/opt/conda/lib/python3.7/site-packages/torch/testing/_internal/common_device_type.py", line 391, in instantiated_test
    raise rte
  File "/opt/conda/lib/python3.7/site-packages/torch/testing/_internal/common_device_type.py", line 378, in instantiated_test
    result = test(self, **param_kwargs)
  File "/opt/conda/lib/python3.7/site-packages/torch/testing/_internal/common_device_type.py", line 828, in test_wrapper
    return test(*args, **kwargs)
  File "functorch/test_vmap.py", line 3509, in test_op_has_batch_rule
    skip_inplace=inplace_failures)
  File "functorch/test_vmap.py", line 3191, in opinfo_vmap_test
    check_vmap_fallback(self, test, op)
  File "/var/lib/jenkins/workspace/test/functorch/common_utils.py", line 427, in check_vmap_fallback
    thunk()
  File "functorch/test_vmap.py", line 3182, in test
    self.vmap_outplace_test(func, args, kwargs, in_dims, check_shape_only, postprocess_fn)
  File "functorch/test_vmap.py", line 3129, in vmap_outplace_test
    for loop_out, vmap_out in compute_quantities_for_vmap_test(func, args, kwargs, in_dims):
  File "/var/lib/jenkins/workspace/test/functorch/common_utils.py", line 284, in compute_quantities_for_vmap_test
    out_dim, batch_size, compute_loop_out, clone_inputs):
  File "/var/lib/jenkins/workspace/test/functorch/common_utils.py", line 253, in _compute_quantities_for_vmap_test
    batched_out = vmap(op, in_dims=in_dims, out_dims=out_dim)(*batched_args, **kwarg_values)
  File "/opt/conda/lib/python3.7/site-packages/functorch/_src/vmap.py", line 363, in wrapped
    func, batch_size, flat_in_dims, flat_args, args_spec, out_dims, randomness, **kwargs
  File "/opt/conda/lib/python3.7/site-packages/functorch/_src/vmap.py", line 35, in fn
    return f(*args, **kwargs)
  File "/opt/conda/lib/python3.7/site-packages/functorch/_src/vmap.py", line 489, in _flat_vmap
    batched_outputs = func(*batched_inputs, **kwargs)
RuntimeError: aten::unfold_copy hit the vmap fallback which is currently disabled
```
@zou3519 what's the correct way to fix this?

File: aten/src/ATen/functorch/BatchRulesDecompositions.cpp (hunk 4 above)
Comment (zou3519) [reply aten/src/ATen/functorch/BatchRulesDecompositions.cpp:264]: Thanks for catching that

**Comment by zou3519** [comment]
> @zou3519 what's the correct way to fix this?

This PR removed the batching rule for unfold_copy, the right fix is to add a batching rule for it. Likely looks something similar to the following (pseudocode), I am not sure what the types of unfold_copy's args are.
```
std::tuple<Tensor, optional<int64_t>> unfold_copy_batch_rule(
    const Tensor &self, optional<int64_t> self_bdim, int64_t dim, SymInt size, int64_t step)
{
  TORCH_INTERNAL_ASSERT(self_bdim.has_value());
  auto self_ = moveBatchDimToFront(self, self_bdim);
  auto logical_rank = rankWithoutBatchDim(self, self_bdim);
  dim = maybe_wrap_dim(dim, logical_rank) + 1;
  if (logical_rank==0) {
    self_ = self_.unsqueeze(-1);
  }
  auto result = self_.unfold_copy(dim, size, step);
  if (logical_rank==0) {
    result = result.squeeze(-1);
  }
  return std::make_tuple(result, 0);
}
```

**Pull request closed without merging**